velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,51 +970,238 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
+
+      // Make sure the escaped char is valid.
+      cursor++;
+      auto current = *cursor;
+      VELOX_USER_CHECK(
+          current == escapeChar || current == '_' || current == '%',
+          "Escape character must be followed by '%', '_' or the escape character itself");
+
+      // Append the escaped char.
+      os << current;
+    } else {
+      // Escape char not found, append all the non-escape chars.
+      appendChars(previous, endCursor);
+      break;
+    }
+
+    // Advance the cursor.
+    cursor++;
+  }
+
+  return os.str();
+}
+
+/// An iterator that provides methods(hasNext, next) to iterate through a
+/// pattern string. We call hasNext to see if there are more chars, call next
+/// to advance the cursor to next char.
+class PatternStringIterator {
+ public:
+  /// Represents the state of current cursor/char.
+  enum class CharKind {
+    // Escape char.
+    // NOTE: If escape char is set as '\',  for pattern '\\', the first '\' is
+    // an escaping char, the second is not, it is just a literal '\'
+    kEscape,
+    // Wildcard char.
+    // NOTE: If escape char is set as '\', for pattern '\%%', the first '%' is
+    // not a wildcard, just a literal '%', the second '%' is a wildcard.
+    kWildcard,
+    // Chars that are not escape char & not wildcard char.
+    kNormal
+  };
+
+  PatternStringIterator(StringView pattern, std::optional<char> escapeChar)
+      : pattern_(pattern), escapeChar_(escapeChar) {}
+
+  /// Are there more chars to iterate?
+  bool hasNext() const {
+    return currentIndex_ < (int32_t)(pattern_.size() - 1);
+  }
+
+  /// Advance the cursor to next char, escape char is automatically handled.
+  bool next() {
+    nextInternal();
+    userPreviousCharKind_ = previousCharKind_;
+
+    if (charKind_ == CharKind::kEscape) {
+      // Escape char should be followed by another char.
+      if (!hasNext()) {
+        return false;
+      }
+
+      // Advance the cursor.
+      nextInternal();
+
+      auto currentChar = current();
+      // The char follows escapeChar can only be one of (%, _, escapeChar).
+      if (currentChar != '%' && currentChar != '_' &&
+          currentChar != escapeChar_) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /// Current index of the cursor.
+  char currentIndex() const {
+    return currentIndex_;
+  }
+
+  /// Char at current cursor.
+  char current() const {
+    return pattern_.data()[currentIndex_];
+  }
+
+  /// Kind of the current char.
+  CharKind charKind() const {
+    return charKind_;
+  }
+
+  /// Kind of previous char(escape char is skipped).
+  CharKind previousCharKind() const {
+    return userPreviousCharKind_;
+  }
+
+ private:
+  const StringView pattern_;
+  const std::optional<char> escapeChar_;
+
+  int32_t currentIndex_ = -1;
+  // Kind of current char.
+  CharKind charKind_ = CharKind::kNormal;
+  // Kind of previous char(literally).
+  CharKind previousCharKind_ = CharKind::kNormal;
```

Comment:
I think we can remove this member variable. See https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:optimize_like_when_user_specified_escape_char?expand=1

Reply:
The link seems broken, we need `previousCharKind_` to track escape state?

Reply:
@mbasmanova Applied your change.